onnxruntime/core/graph/function_utils.cc

```diff
@@ -114,17 +114,29 @@ static void IOTypeConstraintHelper(const ONNX_NAMESPACE::FunctionProto& onnx_fun
 
   std::function<void(const ONNX_NAMESPACE::NodeProto&)> process_node = [&](const ONNX_NAMESPACE::NodeProto& node) {
     auto it = opset_imports.find(node.domain());
-    ORT_ENFORCE(it != opset_imports.end(), "No opset registered for domain " + node.domain() + " in function opset imports.");
+    ORT_ENFORCE(it != opset_imports.end(),
+                "No opset registered for domain " + node.domain() + " in function opset imports.");
     int domain_version = it->second;
-    ORT_ENFORCE(domain_version != -1, "No opset registered for domain " + node.domain() + " in function opset imports.");
-    const auto node_op_schema =
-        schema_registry->GetSchema(node.op_type(), domain_version, node.domain());
+    ORT_ENFORCE(domain_version != -1,
+                "No opset registered for domain " + node.domain() + " in function opset imports.");
+
+    const auto* node_op_schema = schema_registry->GetSchema(node.op_type(), domain_version, node.domain());
+    int variadic_arg_idx = -1;
     for (int i = 0; i < node.input_size(); ++i) {
       auto& in_name = node.input().Get(i);
       auto iter = input_name_idx_map.find(in_name);
       if (iter != input_name_idx_map.end()) {
         int idx = iter->second;
-        std::string type_str = node_op_schema ? node_op_schema->inputs().at(i).GetTypeStr() + "in" + std::to_string(idx) : "Tin" + std::to_string(idx);
+        // if we have hit a variadic arg it is the last input in the schema, so we need to use that index not i.
+        auto schema_idx = variadic_arg_idx != -1 ? variadic_arg_idx : i;
+
+        std::string type_str;
+        if (node_op_schema) {
+          type_str = node_op_schema->inputs().at(schema_idx).GetTypeStr() + "in" + std::to_string(schema_idx);
```

Comment:
I suggest `std::to_string(idx)` instead of `std::to_string(schema_idx)`. The actual name used does not matter as much as ensuring the uniqueness of the names. In the variadic case with variadic inputs X and Y, it is better to generate different type names for X and Y ... this doesn't matter for Concat where all are expected to have same type, but variadic inputs/outputs of If/Loop/Scan etc. are allowed to have different types.

Reply:
Isn't it just going to read the same list of types for each one though given it's reading them from schema_idx below? i.e. you'll generate the same constraint list but with a different name.

Reply:
Yes. But a schema that says X and Y have the same type T, where T is in some set {...}, is stricter than a schema that says X has type T1 in set {...} and Y has type T2 in set {...}  (even though the sets are the same). In the first case it is expected that X and Y will have the same type, and in the second case this is not required.